WSearchLineEdit: Add options to disable auto-completion and history

fwcd opened the pull request:
### Fixes #10634

This adds two options (that are enabled by default) to configure whether search completions and history (including up/down keyboard shortcuts) should be enabled:

<img width="1100" alt="image" src="https://user-images.githubusercontent.com/30873659/194570181-d759564e-70bd-499d-a74a-a58828772360.png">

fwcd committed:
DlgPrefInterface: Add checkboxes for configuring search bar

fwcd committed:
DlgPrefInterface: Wire up search bar options

fwcd committed:
WSearchLineEdit: Respect history enabled setting

fwcd committed:
WSearchLineEdit: Respect search completion option

fwcd commented:
Hm, the build failure is strange, is `setAutoCompletion` not available on all platforms?

fwcd commented:
It's mostly minor things that I find a bit inconvenient:

- The completions kicking in too quickly, as you've described in the issue
- Accidentally hitting up/down or turning the track selection knob on the controller now moves through the history. While this of course is the intended behavior, I find it a bit irritating since I now have to seek back through the history and figure out what I previously wanted to type
- I don't really use the feature, so being able to disable it makes the UI a bit less noisy

ronso0 commented:
Just a side note, the new options should be in the Library page.

ronso0 commented:
> Figure out why setAutoCompletion doesn't compile on all platforms

Use [`setCompleter(completer*)`](https://doc.qt.io/qt-5/qcombobox.html#setCompleter)?
Either `completer*` is a valid QCompleter or a nullptr (disables auto-completion as described in the doc).

Also, instead of parsing the config on each Up/Down/Space press, why not use a static functions to set bools, like the one setting the debounce timeout?
[void WSearchLineEdit::setDebouncingTimeoutMillis(int debouncingTimeoutMillis)](https://github.com/mixxxdj/mixxx/blob/49164318d384c1c92d3bf73cac527c1af50af289/src/widget/wsearchlineedit.cpp#L63-L66)

Furthermore, I'm not comfortable with removing the history function altogether (which is not the case now btw, queries are still saved and restored, though not accessible).
Isn't it sufficient for your use case to suppress the emulated keystrokes [sent by librarycontrol](https://github.com/mixxxdj/mixxx/blob/49164318d384c1c92d3bf73cac527c1af50af289/src/library/librarycontrol.cpp#L664-L680), i.e. still allow regular hotkeys (Up, Down, Ctrl+Space) and keep the expand button?

fwcd committed:
DlgPrefLibrary: Move search history options here

fwcd committed:
Library: Rename option to EnableSearchHistoryShortcuts

fwcd committed:
WSearchLineEdit: Use setCompleter to set completions

fwcd commented:
I've moved the settings as per your suggestion to the Library page (see the updated screenshot in the PR description) and renamed the history setting to 'Enable search history shortcuts' to clarify that it only affects the shortcuts and not the history itself.

Suppressing the emulated keystrokes in `LibraryControl` would probably work but IMO feels a bit like the wrong place to model this (if we consider the search history to be an implementation detail of the search bar, we should also handle it there and not swallow the key events at a higher level). On the other hand, the generated keys are already somewhat tied to the search bar's semantics. Not sure about this one.

Regarding the static functions: While this may work for the history shortcuts (since we only need to read the value during key events), `setSearchCompletionsEnabled` cannot really be static if we want to to swap the completer based on whether search completions are enabled. The todo comment mentions that this also is more of a workaround currently and that it would be better to wire up the config changes with signals and slots. Since we would have access to the `WSearchLineEdit` instance that way, this would also solve the aforementioned issue.

Since I'm not that familiar with the data flow of configuration changes yet, do you have some pointers on how to do cleanly propagate config changes to `WSearchLineEdit`?

ronso0 commented:
> `setSearchCompletionsEnabled` cannot really be static if we want to to swap the completer based on whether search completions are enabled.

You could check if this bool changed in the focusInEvent()

ronso0 commented:
Anyways, I assume this setting is changed rarely, so applying this after start / skin change/reload may be sufficient IMO

fwcd committed:
WSearchLineEdit: Model search history options as statics

As per the PR suggestions:

- https://github.com/mixxxdj/mixxx/pull/10942#issuecomment-1270717206
- https://github.com/mixxxdj/mixxx/pull/10942#issuecomment-1272291421

fwcd committed:
WSearchLineEdit: Use custom QCompleter

fwcd commented:
I've now wired up the settings as static bools as you suggested. Also, `WSearchLineEdit` now creates a custom owned `QCompleter` that is set depending on whether search completions are enabled.

One thing that struck me as slightly odd is that some settings are updated differently: For example, the debounce time is updated immediately when the user changes the widget whereas most other settings are first applied when the user clicks 'Apply'. This feels a bit inconsistent (but is probably a topic for another PR). With these new settings, I've tried to match the original semantics (i.e. first update the config when 'Apply' is clicked).

In the longer term, we'd probably want to find a clean way to notify `WSearchLineEdit` about config changes, the updating-static-variables feels like a workaround to a deeper architectural issue currently.

ywwg commented:
I support this feature, I do not personally like the autocomplete / history but I understand others may find it useful.

ronso0 commented on src/widget/wsearchlineedit.cpp:
```diff
@@ -666,6 +696,16 @@ void WSearchLineEdit::updateClearAndDropdownButton(const QString& text) {
+#if ENABLE_TRACE_LOG
+    kLogger.trace()
+            << "updateCompleter"
+            << text;
```
undefined, and not useful here

ronso0 commented on src/widget/wsearchlineedit.h:
```diff
@@ -1,5 +1,7 @@
 #pragma once
 
+#include <qcompleter.h>
```
why not `<QCompleter>` like the other Qt includes?

fwcd replied:
This was what clangd auto-included it under, but I agree that `<QCompleter>` is cleaner and more consistent

ronso0 commented on src/widget/wsearchlineedit.h:
```diff
@@ -74,12 +80,15 @@ class WSearchLineEdit : public QComboBox, public WBaseWidget {
     // value provider that sends signals whenever the corresponding
     // configuration value changes.
     static int s_debouncingTimeoutMillis;
+    static bool s_searchCompletionsEnabled;
+    static bool s_searchHistoryShortcutsEnabled;
```
```suggestion
    static bool s_searchCompletionsEnabled;
    static bool s_searchHistoryShortcutsEnabled;
```
just because without 'search' prefix readability in the cpp is better IMO, and less redundant e.g. when getting it from within the preferences `WSearchLineEdit::kEnableCompletionsByDefault`

ronso0 commented on src/preferences/dialog/dlgpreflibrarydlg.ui:
```diff
@@ -361,6 +361,31 @@
+         <string>Enable search history shortcuts</string>
+        </property>
+       </widget>
+      </item>
```
the checkboxes should be added to `<tabstops>` near the end of the file

ronso0 reviewed (changes requested):
I'm sorry you had to wait so long, this was kinda pushed off the table.

It's working nicely, thanks. Just some nitpicks.
I guess it doesn't matter that the completer is always created even if it will never actually be loaded to lineEdit->completer.

ronso0 commented on src/widget/wsearchlineedit.h:
```diff
@@ -22,9 +24,13 @@ class WSearchLineEdit : public QComboBox, public WBaseWidget {
     static constexpr int kMaxDebouncingTimeoutMillis = 9999;
     static constexpr int kSaveTimeoutMillis = 5000;
     static constexpr int kMaxSearchEntries = 50;
+    static constexpr bool kEnableSearchCompletionsByDefault = true;
+    static constexpr bool kEnableSearchHistoryShortcutsByDefault = true;
```
```suggestion
    static constexpr bool kEnableSearchCompletionsByDefault = true;
    static constexpr bool kEnableSearchHistoryShortcutsByDefault = true;
```
because 'search' is redundant e.g. when getting it from within the preferences `WSearchLineEdit::kEnableCompletionsByDefault`

ronso0 commented:
I think it works also with the "Search history:" label (also we need less space and thereby potentially avoid horizontal scrollbars for long translations)
![image](https://user-images.githubusercontent.com/5934199/203182515-79bd3c8a-5a10-46c2-87b0-56242f34e376.png)

- [ ] Enable search history **keyboard** shortcuts ?

fwcd committed:
WSearchLineEdit: Apply PR suggestions

fwcd commented:
I've updated the branch with your suggestions

ronso0 commented on src/widget/wsearchlineedit.cpp:
```diff
@@ -1,14 +1,13 @@
-
 #include <QAbstractItemView>
 #include <QApplication>
+#include <QCompleter>
```
Oh, during my previous review I didn't notice this is already included in wsearchlineedit.h
Please remove it here, git commit --amend (maybe name it more explicitely) and push one last time.

FYI rebasing after a review _without approval_ is not appreciated since that means someone has to review again, instead of just checking the last commits. Here, it's not much of a drama because the change set is rather small and can be verified by manual testing.

fwcd replied:
> FYI rebasing after a review without approval is not appreciated

Ah sorry, I'll try to keep that in mind for the next time

ronso0 replied:
I didn't neither review again nor test manually, so if you like to --fixup previous commits with the latest changes and squash..  feel free.

fwcd replied:
I went with amending for now, most of these are just small stylistic changes anyway

fwcd commented:
Any updates on this?

daschuer commented:
After https://github.com/mixxxdj/mixxx/pull/11207 is merged, the completion is no longer applied immediately. It requires "→" or "⏎" to take effect. 

This means the 

> The completions kicking in too quickly, as you've described in the issue

part will be solved. 

Can we reduce this to only add the "EnableSearchHistoryShortcuts" preferences option for less clutter in the preferences?

daschuer commented on src/widget/wsearchlineedit.h:
```diff
@@ -22,9 +23,13 @@ class WSearchLineEdit : public QComboBox, public WBaseWidget {
     static constexpr int kMaxDebouncingTimeoutMillis = 9999;
     static constexpr int kSaveTimeoutMillis = 5000;
     static constexpr int kMaxSearchEntries = 50;
+    static constexpr bool kEnableCompletionsByDefault = true;
+    static constexpr bool kEnableHistoryShortcutsByDefault = true;
```
This is a function name, how about 
```suggestion
    static constexpr bool kDefaultSearchCompletionsEnabled = true;
    static constexpr bool kDefaultSearchHistoryShortcutsEnabled = true;
```

ronso0 replied:
[I preferred to remove](https://github.com/mixxxdj/mixxx/pull/10942#pullrequestreview-1189077067) the redundant 'search' string, so let's name these
`kCompletionsEnabledDefault`
`kHistoryShortcutsEnabledDefault`
?

fwcd commented:
I still think having the option to disable auto-completion would be cool, if only to reduce visual noise

daschuer commented on src/preferences/dialog/dlgpreflibrary.cpp:
```diff
@@ -495,6 +513,15 @@ void DlgPrefLibrary::slotSearchDebouncingTimeoutMillisChanged(int searchDebounci
     WSearchLineEdit::setDebouncingTimeoutMillis(searchDebouncingTimeoutMillis);
 }
 
+void DlgPrefLibrary::searchHistoryOptionsChanged() {
```
Since this is not a slot, name the function according to what it does, not when it is called. 
```suggestion
void DlgPrefLibrary::updateSearchLineEditHistoryOptions() {
```

daschuer commented:
> I still think having the option to disable auto-completion would be cool, if only to reduce visual noise

OK

fwcd committed:
WSearchLineEdit: Address PR suggestions

daschuer reviewed (approved):
LGTM Thank you.
